auth/auth.go

```diff
@@ -35,79 +74,211 @@ func New(authURL string, insecure bool) (client *Client, err error) {
+	return streamer(ctx, desc, cc, method, opts...)
+}
+
+// Authenticate makes a request to the Quarterdeck server with the available API keys
+// in order to fetch new access and refresh tokens. The tokens are returned directly.
+func (c *Client) Authenticate(ctx context.Context, apikey *APIKey) (tokens *Tokens, err error) {
+	// Check to make sure we have APIKeys in order to perform the authentication.
+	if apikey == nil || apikey.ClientID == "" || apikey.ClientSecret == "" {
+		return nil, ErrNoAPIKeys
+	}
 
 	var req *http.Request
-	if req, err = c.NewRequest(ctx, http.MethodPost, "/v1/authenticate", c.apikey); err != nil {
-		return err
+	if req, err = c.newRequest(ctx, http.MethodPost, "/v1/authenticate", apikey); err != nil {
+		return nil, err
 	}
 
-	c.tokens = &Tokens{}
-	if _, err = c.Do(req, c.tokens); err != nil {
-		return err
+	tokens = &Tokens{}
+	if _, err = c.do(req, tokens); err != nil {
+		return nil, err
 	}
 
-	return nil
+	return tokens, nil
 }
 
-func (c *Client) Refresh(ctx context.Context) (err error) {
-	// TODO: check refresh token is available
-	// TODO: check refresh token to make sure it is not expired
+// Refresh makes a request to the Quarterdeck server with the refresh token in order to
+// fetch a new access token. If the refresh token is expired an error is returned. The
+// new tokens are returned directly.
+func (c *Client) Refresh(ctx context.Context, refresh *Tokens) (tokens *Tokens, err error) {
+	var req *http.Request
+	if req, err = c.newRequest(ctx, http.MethodPost, "/v1/refresh", refresh); err != nil {
+		return nil, err
+	}
 
-	tokens := &Tokens{
-		RefreshToken: c.tokens.RefreshToken,
+	tokens = &Tokens{}
+	if _, err = c.do(req, tokens); err != nil {
+		return nil, err
 	}
 
+	return tokens, nil
+}
+
+// Status makes a request to the Quarterdeck server to check if the service is online
+// and ready to make requests. The status check is returned directly.
+func (c *Client) Status(ctx context.Context) (status *Status, err error) {
 	var req *http.Request
-	if req, err = c.NewRequest(ctx, http.MethodPost, "/v1/refresh", tokens); err != nil {
-		return err
+	if req, err = c.newRequest(ctx, http.MethodPost, "/v1/status", nil); err != nil {
+		return nil, err
 	}
 
-	c.tokens = &Tokens{}
-	if _, err = c.Do(req, c.tokens); err != nil {
+	status = &Status{}
+	if _, err = c.do(req, status); err != nil {
+		return nil, err
+	}
+
+	return status, nil
+}
+
+// Wait for ready polls the Quarterdeck status endpoint until it responds with a 200,
+// retrying with exponential backoff or until the context deadline is expired. If the
+// input context does not have a deadline, then a default deadline of 5 minutes is used
+// so this method does not block indefinitely. When the Quarterdeck service is ready
+// then no error is returned; if the Quartdeck does not respond within the retry window
+// an error is returned.
+func (c *Client) WaitForReady(ctx context.Context) (err error) {
```

Comment:
I was going to have this in the `Login` method but then decided not to; we can do retries in the future though so I just left the code in.